thcrap_tsa/src/anm.cpp

```diff
@@ -1054,18 +1318,104 @@ int patch_anm(void *file_inout, size_t size_out, size_t size_in, const char *fn,
 			for(const auto &sprite : entry.sprites) {
 				bounds_draw_rect(bounds, entry.x, entry.y, sprite);
 			}
-			// Do the patching
-			stack_game_png_apply(entry);
+
+			if (entry.thtxoffset) {
+				// Do the patching
+				thtx_header_t* thtx_in = (thtx_header_t*)(anm_entry_in + entry.thtxoffset);
+
+				img_patch_t img_patch = img_get_patch(entry, thtx_in);
+				stack_game_png_apply(img_patch, entry.sprites);
+				img_encode(img_patch);
+
+				thtx_header_t* thtx_out = (thtx_header_t*)(anm_entry_out + entry.thtxoffset);
+				thtx_out->size = img_patch.img_size;
+
+				memcpy(thtx_out->data, img_patch.img_data, img_patch.img_size);
+
+				uint32_t* nextoffset_ptr = anm_get_nextoffset_ptr(anm_entry_out);
+				size_t out_nextoffset = entry.thtxoffset + sizeof(thtx_header_t) + img_patch.img_size;
+				if (*nextoffset_ptr)
+					*nextoffset_ptr = out_nextoffset;
+				anm_entry_out += out_nextoffset;
+
+				free(img_patch.img_data);
+			}
+			else {
+				goto next_no_thtx;
+			}
+		}
+		else if (entry.next) {
+next_no_thtx:
+			anm_entry_out += entry.next - anm_entry_in;
 		}
 		if(!entry.next) {
 			bounds_store(name_prev, bounds);
 		}
-		anm_entry_out = entry.next;
+		anm_entry_in = entry.next;
 		anm_entry_clear(entry);
 	}
 	png_image_clear(bounds);
 	png_image_clear(png);
+	free(anm_entry_in);
 
 	log_debugf("-------------\n");
 	return 1;
 }
+
+typedef uint8_t* __fastcall file_load_t(const char* fn, size_t& out_size, bool from_archive);
+
+size_t anm_get_size(const char* fn, json_t* patch, size_t patch_size) {
+	tlnote_remove();
+
+	if (game_id < TH19) {
+		return 0;
+	}
+
+	uintptr_t file_load = (CurrentImageBase + 0xC82A0);
+
+	// Hack
+	file_rep_t* fr = fr_tls_get();
+	fr->disable = true;
+
+	size_t out_size;
+	size_t* pout_size = &out_size;
+	void* anm_file;
+	anm_header11_t* entry;
+
+	// WARNING: INLINE ASSEMBLY
+	// TH19's file_load function uses a calling convention that's basically __fastcall...
+	// except that the caller has to do the stack cleanup
+	__asm {
+		push 0
+		mov ecx, fn
+		mov edx, pout_size
+		mov eax, file_load
+		call eax
+		mov anm_file, eax
+		add esp, 4
+	}
```

Comment:
how about making a wrapper in a codecave with the correct calling convention? something like this
```
ff 74 24 04    push dword [esp+4]
b8 a0 82 4c 00 mov eax,0x4c82a0
ff d0          call eax
83 c4 04       add esp,4
c2 04 00       ret 4
```